### PR TITLE
fixed Automation-io payload types

### PR DESCRIPTION
**Pull request by dakhnod**
This PR fixes the types of the Automation-IO characteristic payload.

While each pin indeed occupies two bits, multiple pins will occupy multiple pairs in one byte.

### CODE REVIEW

**Comment by shmuelzon** [comment]
Hey,

I'm wondering if this is the correct way to go. If the 2-bit type is [defined](https://www.bluetooth.com/specifications/specs/automation-io-service-1-0/) as an array of 2 bit values, shouldn't we just parse it as such? That is, convert whatever's left of the buffer to comma-separated values for each 2 bits in `chartoa()`? This is also the only characteristic I see that's defined with the 2-bit type so I think we should be OK.
As for the `Digital Output` characteristic you changed, it's already uint8, so what's the point of removing the definition? Unless it's supposed to be something else entirely...

**Comment by dakhnod** [comment]
If I understand correctly, if you define the type as uint8 or uint2, the first n bits are parsed as that and the rest is considered unknown. So, what we could do is introduce a new type, `digital2` or something, that consumes all repeating tuplets that are not `0b11`, since `0b11` means undefined. So something like `0b11110100` would result in `1,0`. What would you do if we receive `0b11110011`? Just put out `0`? Then we would not know that it's the second pin. Somehow indicate that pin 0 is undefined? `0, u`? `0, x`? `0, -`?

**Comment by dakhnod** [comment]
If I understand correctly, if you define the type as uint8 or uint2, the first n bits are parsed as that and the rest is considered unknown. So, what we could do is introduce a new type, `digital2` or something, that consumes all repeating tuplets that are not `0b11`, since `0b11` means undefined. So something like `0b11110100` would result in `1,0`. What would you do if we receive `0b11110011`? Just put out `0`? Then we would not know that it's the second pin. Somehow indicate that pin 0 is undefined? `0, u`? `0, x`? `0, -`?

**Comment by shmuelzon** [comment]
The fact that `0b11` is undefined is unique to that specific characteristic and not to the `2bit` type.
Just like `utf8` consumes the entire buffer we can also have `2bit` do the same. However, it won't have any special characters as all it can do is convert the bits to decimal and publish them as comma-separated values. Undefined will simply be `3`, in this case.

As for unknown data types, if we have more data than defined types and we don't know what to do with the rest of the data, it's simply regarded as `uint8`. Hence your change to `Digital Output` doesn't do anything.

**Comment by dakhnod** [comment]
> The fact that `0b11` is undefined is unique to that specific characteristic and not to the `2bit` type. Just like `utf8` consumes the entire buffer we can also have `2bit` do the same. However, it won't have any special characters as all it can do is convert the bits to decimal and publish them as comma-separated values. Undefined will simply be `3`, in this case.
> 
> As for unknown data types, if we have more data than defined types and we don't know what to do with the rest of the data, it's simply regarded as `uint8`. Hence your change to `Digital Output` doesn't do anything.

I would still suggest to not use `2bit` for that, since there might be a future usecase where you will need to use `2bit` and it will not be repeating, aka you'll need to discard the remaining 6 bits. Not sure about the naming though, maybe something like `2bit_aio´? Also, should we discard trailing `undefined`? Something like `0b11110001` would just become '0, 1' instead of '3, 3, 0, 1'.

**Comment by dakhnod** [comment]
Anyway, wdyt about my proposal?

**Comment by shmuelzon** [comment]
I don't know it it would make sense to discard the rest of the byte. One could have just defined the characteristic as a byte and be done with it. The only reason you'd define a 2-bit value, is if you want to make use of the rest of the byte. What may happen is that a characteristic can be defined, for example, as 2-bit, 2-bit, 4-bit, or any other separation of bits. This would be even more problematic for us to handle since then we'd need to remember the offset inside each byte. Also, we'd need to define this specific characteristic as
```c
static characteristic_type_t types_00002a57_0000_1000_8000_00805f9b34fb[] = { CHAR_TYPE_2BIT, CHAR_TYPE_2BIT, CHAR_TYPE_2BIT, CHAR_TYPE_2BIT, -1 };
```
In the hopes that there aren't more than 4 inputs... Otherwise, the definition of this characteristic will be device specific.
There are also characteristic types of 4-bit and 12-bit. We can try to see which characteristics use those and how they are expected to be used and, if they all follow the same logic, we can do that.

File: main/ble_utils.c
Comment (shmuelzon) [inline main/ble_utils.c:303]: Why are you printing from the MSB first? Isn't the LSB the first digital IO?

File: main/ble_utils.c
Comment (shmuelzon) [inline main/ble_utils.c:302]: Please make sure to follow the current coding style in terms of spaces and defining variables in the for loop.

**Review by shmuelzon — CHANGES_REQUESTED** [review]

File: main/ble_utils.c
Comment (dakhnod) [reply main/ble_utils.c:303]: reversing the order could yield something like "0, 1, 1, 0, 3, 3, 0, 1" for 6 defined IOs, which I wanted to avoid.

File: main/ble_utils.c
Comment (shmuelzon) [reply main/ble_utils.c:302]: Is this from your implementation of this characteristic or a peripheral that passed the Bluetooth SIG certification?
I'd expect the LSB to be used first, especially since everything else in BLE is little endian

File: main/ble_utils.c
Comment (shmuelzon) [reply main/ble_utils.c:303]: Is this from your implementation of this characteristic or a peripheral that passed the Bluetooth SIG certification?
I'd expect the LSB to be used first, especially since everything else in BLE is little endian

Comment (dakhnod) [reply main/ble_utils.c:303]: Mine. And maybe I am wrong. I am just not aware of any other implementation out there.

Comment (shmuelzon) [reply main/ble_utils.c:303]: I'm not sure if I'm understanding it correctly either, but here's the spec: https://www.bluetooth.org/DocMan/handlers/DownloadDoc.ashx?doc_id=304972#page10

> The Digital characteristic contains the values of these bits in little endian order

Comment (dakhnod) [reply main/ble_utils.c:303]: Little endian does not mean LSB or MSB

Comment (shmuelzon) [reply main/ble_utils.c:303]: True, but it doesn't make sense, to me, to read the bytes in little endian but then, to read the 2-bits from MSB to LSB but, read the 2-bit values in little endian again.
Also, since it's defined as a 2-bit array, I'd expect the array elements to follow each other in memory instead of jumping back and forth.

Comment (dakhnod) [reply main/ble_utils.c:303]: I followed the natural reading direction, which  is left to right for me. If you insist we can change the direction though.

Comment (dakhnod) [reply main/ble_utils.c:303]: Okay, it's LSB now

**Review by shmuelzon — APPROVED** [review]

**Comment by shmuelzon** [comment]
Hey, I pushed a commit to your branch to align the coding style. I also removed the `fill up remaining bits with 0b11` section as I feel it's up to the application to know to do that, especially since the `0b11` value is specific to the DigitalIO characteristic.
Can you please make sure it still works on your end and, once confirmed, I'll merge it.

Thanks!

**Comment by dakhnod** [comment]
I find it important to fill them up. Let's create a new Datatype specifically for AIO then.

**Comment by shmuelzon** [comment]
That's not something I'd be comfortable with adding. The `0b11` is very application specific and the application needs to be aware of stray `0b11`s as it would get them from the MQTT message. It doesn't seem like a stretch, to me, to make sure it appends them when it wants to set.
I'd be happy to merge this as-is to the project but if you really need to add the `0b11`s, I suggest doing it as a part of a fork.

**Comment by dakhnod** [comment]
Sorry, but I find the padding to be essential for this usecase. Here's my suggestions: we introduce a new type 2bit_padding_1 or something and use it here. That would pad the bytes with ones then, and is not too application specific.

**Comment by shmuelzon** [comment]
But then the type is application specific. Can't you just add the padding in whatever application you're using to read/write from the MQTT bus? It's going to need to remove the padding when it reads anyway...

**Comment by dakhnod** [comment]
Sure, if you insist. Works on my side.

**Comment by shmuelzon** [comment]
Merged, thanks!

**Pull request merged**
